src/pybind/mgr/cephadm/module.py

```diff
@@ -726,6 +727,21 @@ def __init__(self, *args, **kwargs):
         # in-memory only.
         self.offline_hosts: Set[str] = set()
 
+        # services:
+        self.osd_service = OSDService(self)
+        self.nfs_service = NFSService(self)
+        self.mon_service = MonService(self)
+        self.mgr_service = MgrService(self)
+        self.mds_service = MdsService(self)
+        self.rgw_service = RgwService(self)
+        self.rbd_mirror_service = RbdMirrorService(self)
+        self.grafana_service = GrafanaService(self)
+        self.alertmanager_service = AlertmanagerService(self)
+        self.prometheus_service = PrometheusService(self)
+        self.node_exporter_service = NodeExporterService(self)
+        self.crash_service = CrashService(self)
+        self.iscsi_servcie = IscsiService(self)
```

Comment:
Instead of pre-allocating .. would it be better to initialize these when they are actually used?

Reply:
would be a follow-up PR. Especially in conjunction with `_apply_service`